### ISSUE TITLE
Proposal: Make it easy to setup Width and Height on CommandBarFlyout

### ISSUE THREAD

**Issue by zhuxb711** [microsoft/microsoft-ui-xaml#4301]
# Proposal: Make it easy to setup Width and Height on CommandBarFlyout 

## Summary

I could not custom the (MaxHeight) height or (MaxWidth) width on commandbarflyout.
Please make it easy.

## Rationale

* Developer could be easier to custom Commandbarflyout
* Sometimes CommandBarFlyout will extend out of the window if no limit on MaxHeight

## Scope

| Capability | Priority |
| :---------- | :------- |
| Add Height | Must |
| Add Width| Must |
| Add MaxHeight | Should |
| Add MaxWidth | Should |

**Comment by ranjeshj** [comment]
@YuliKl @adambarlow I'm not sure if there was some motivation for doing it this way, but the presenter is hardcoded to have [infinite max height](https://github.com/microsoft/microsoft-ui-xaml/blob/dfbe38bfb2996dfaa2f09544c4e86aa69ccc906b/dev/CommandBarFlyout/CommandBarFlyout.cpp#L237).

**Comment by zhuxb711** [comment]
@ranjeshj Yes. So if there are too many things in CommandBarFlyout, it will be out of my Monitor

**Comment by adambarlow** [comment]
I think @llongley might have an idea on this one. 

@zhuxb711 how many items causes it to look incorrect?

**Comment by mdtauk** [comment]
Is there any behaviour built in to allow a carosel style scrolling for the secondary items in a CommandBarFlyout - for when there is not enough space to display all items?

![image](https://user-images.githubusercontent.com/7389110/109044373-3f282900-76ca-11eb-8090-722d20972ead.png)

I don't see any tests for such behaviour in the TestApp

**Comment by zhuxb711** [comment]
Out of monitor
![image](https://user-images.githubusercontent.com/42038175/109046597-e2e7f800-770f-11eb-98f9-2d1a7bc2db3a.png)

@adambarlow About 15+ items

By the way, if CommandBarFlyout.SecondaryCommand have many items and we scroll the CommandBarFlyout.SecondaryCommand from top to bottom slowly, some word might be clipped (Up and down)

**Comment by adambarlow** [comment]
@zhuxb711 , it looks like perhaps the opening logic is what needs to be addressed here. I think it shouldn't open off the monitor, I thought we had logic so that it would take the edge of the screen into account. If not, that is a bug and should be filed as an issue report, IMO.

@mdtauk in the UI guidelines we mention that a commandbar flyout should be used for a limited set of items. Having 15 items in a context menu is very bad UI design. There should be at most 10ish but ideally something like 5.

**Comment by mdtauk** [comment]
> @zhuxb711 , it looks like perhaps the opening logic is what needs to be addressed here. I think it shouldn't open off the monitor, I thought we had logic so that it would take the edge of the screen into account. If not, that is a bug and should be filed as an issue report, IMO.
> 
> @mdtauk in the UI guidelines we mention that a commandbar flyout should be used for a limited set of items. Having 15 items in a context menu is very bad UI design. There should be at most 10ish but ideally something like 5.

That supposes the developer/designer pays attention, and they are not dynamically loading items, where they are unable to control the source of those items.

Just wondering if there is a place for adding logic.  Can secondary items open upwards perhaps, when the control is lower down on the screen?

There is scrolling logic there according to the "Show tall CommandBarFlyout" test in the MUXControlsTestApp - so that is one good thing

![image](https://user-images.githubusercontent.com/7389110/109068308-9688c200-76e7-11eb-934b-749e7814deca.png)

**Comment by zhuxb711** [comment]
A better option is to provide an overflow menu if there are too many items in SecondaryCommand, just like the commandbar

**Comment by mdtauk** [comment]
> A better option is to provide an overflow menu if there are too many items in SecondaryCommand, just like the commandbar

That is already there.  You get the toolbar icons, and a "more" button that displays a drop down menu of additional items.

What we are talking about, is what happens when that bottom menu doesn't fit on the screen.

My suggestion is to enable the menu to appear above the icons, if there is not enough space below.  And the proposal is to enable more specific control of the size of the various parts of the control.

**Comment by roxk** [comment]
Cannot reproduce. The command bar flyout would be shown upwards when the menu target is at the bottom of the screen/xaml root. Could you provide a minimal repro? Even when the menu is opening downwards, there is a scrollbar inside the menu by default when there are many items.
